File: PFAnalisisAlgoritmosGo/algoritmos/StrassenNaivStep.go
package algoritmos

// StrassenNaivStep realiza la multiplicación de matrices utilizando el algoritmo de Strassen
func StrassenNaivStep(A, B [][]float64, cantidadFilasMatrices, m int) [][]float64 {
	Result := make([][]float64, cantidadFilasMatrices)
	for i := range Result {
		Result[i] = make([]float64, cantidadFilasMatrices)
	}

	// Caso base
	if cantidadFilasMatrices <= m {
		return NaivStandard(A, B, cantidadFilasMatrices, cantidadFilasMatrices, cantidadFilasMatrices)
	}

	nuevoTamanio := cantidadFilasMatrices / 2

	// Inicializamos submatrices para A y B
	A11, A12, A21, A22 := crearSubmatrices(nuevoTamanio)
	B11, B12, B21, B22 := crearSubmatrices(nuevoTamanio)
	Result11, Result12, Result21, Result22 := crearSubmatrices(nuevoTamanio)

	// Inicializamos matrices auxiliares y ayudantes
	ayudante1, _, _, _ := crearSubmatrices(nuevoTamanio)
	ayudante2, _, _, _ := crearSubmatrices(nuevoTamanio)

	auxiliar1, _, _, _ := crearSubmatrices(nuevoTamanio)
	auxiliar2, _, _, _ := crearSubmatrices(nuevoTamanio)
	auxiliar3, _, _, _ := crearSubmatrices(nuevoTamanio)
	auxiliar4, _, _, _ := crearSubmatrices(nuevoTamanio)
	auxiliar5, _, _, _ := crearSubmatrices(nuevoTamanio)
	auxiliar6, _, _, _ := crearSubmatrices(nuevoTamanio)
	auxiliar7, _, _, _ := crearSubmatrices(nuevoTamanio)

	// Llenamos las submatrices de A y B
	llenarSubmatrices(A, B, A11, A12, A21, A22, B11, B12, B21, B22, nuevoTamanio)

	// Computamos las siete multiplicaciones necesarias
	ayudante1 = sumarMatrices(A11, A22)
	ayudante2 = sumarMatrices(B11, B22)
	auxiliar1 = StrassenNaivStep(ayudante1, ayudante2, nuevoTamanio, m)

	ayudante1 = sumarMatrices(A21, A22)
	auxiliar2 = StrassenNaivStep(ayudante1, B11, nuevoTamanio, m)

	ayudante1 = restarMatrices(B12, B22)
	auxiliar3 = StrassenNaivStep(A11, ayudante1, nuevoTamanio, m)

	ayudante1 = restarMatrices(B21, B11)
	auxiliar4 = StrassenNaivStep(A22, ayudante1, nuevoTamanio, m)

	ayudante1 = sumarMatrices(A11, A12)
	auxiliar5 = StrassenNaivStep(ayudante1, B22, nuevoTamanio, m)

	ayudante1 = restarMatrices(A21, A11)
	ayudante2 = sumarMatrices(B11, B12)
	auxiliar6 = StrassenNaivStep(ayudante1, ayudante2, nuevoTamanio, m)

	ayudante1 = restarMatrices(A12, A22)
	ayudante2 = sumarMatrices(B21, B22)
	auxiliar7 = StrassenNaivStep(ayudante1, ayudante2, nuevoTamanio, m)

	// Calculamos las submatrices de Result
	Result11 = sumarMatrices(sumarMatrices(restarMatrices(auxiliar1, auxiliar5), auxiliar7), auxiliar4)
	Result12 = sumarMatrices(auxiliar3, auxiliar5)
	Result21 = sumarMatrices(auxiliar2, auxiliar4)
	Result22 = sumarMatrices(sumarMatrices(restarMatrices(auxiliar1, auxiliar2), auxiliar3), auxiliar6)

	// Combinar los resultados en la matriz final
	combinarResultados(Result, Result11, Result12, Result21, Result22, nuevoTamanio)

	return Result
}

// Funciones de utilidad para crear, sumar, restar y combinar submatrices
func crearSubmatrices(tamanio int) ([][]float64, [][]float64, [][]float64, [][]float64) {
	a, b, c, d := make([][]float64, tamanio), make([][]float64, tamanio), make([][]float64, tamanio), make([][]float64, tamanio)
	for i := range a {
		a[i], b[i], c[i], d[i] = make([]float64, tamanio), make([]float64, tamanio), make([]float64, tamanio), make([]float64, tamanio)
	}
	return a, b, c, d
}

func llenarSubmatrices(A, B, A11, A12, A21, A22, B11, B12, B21, B22 [][]float64, tamanio int) {
	for i := 0; i < tamanio; i++ {
		for j := 0; j < tamanio; j++ {
			A11[i][j] = A[i][j]
			A12[i][j] = A[i][j+tamanio]
			A21[i][j] = A[i+tamanio][j]
			A22[i][j] = A[i+tamanio][j+tamanio]

			B11[i][j] = B[i][j]
			B12[i][j] = B[i][j+tamanio]
			B21[i][j] = B[i+tamanio][j]
			B22[i][j] = B[i+tamanio][j+tamanio]
		}
	}
}

func sumarMatrices(A, B [][]float64) [][]float64 {
	N := len(A)
	C := make([][]float64, N)
	for i := range C {
		C[i] = make([]float64, N)
		for j := 0; j < N; j++ {
			C[i][j] = A[i][j] + B[i][j]
		}
	}
	return C
}

func restarMatrices(A, B [][]float64) [][]float64 {
	N := len(A)
	C := make([][]float64, N)
	for i := range C {
		C[i] = make([]float64, N)
		for j := 0; j < N; j++ {
			C[i][j] = A[i][j] - B[i][j]
		}
	}
	return C
}

func combinarResultados(Result, C11, C12, C21, C22 [][]float64, tamanio int) {
	for i := 0; i < tamanio; i++ {
		for j := 0; j < tamanio; j++ {
			Result[i][j] = C11[i][j]
			Result[i][j+tamanio] = C12[i][j]
			Result[i+tamanio][j] = C21[i][j]
			Result[i+tamanio][j+tamanio] = C22[i][j]
		}
	}
}

File: PFAnalisisAlgoritmosPython/algoritmos/StrassenNaivStep.py
from algoritmos import NaivStandard

def StrassenNaivStep(A, B, Result, cantidadFilasMatrices, m):

    nuevoTamanio = 0
    if (cantidadFilasMatrices % 2 == 0) and (cantidadFilasMatrices > m):
        nuevoTamanio = cantidadFilasMatrices // 2
        
        matrizA11 = []
        matrizA12 = []
        matrizA21 = []
        matrizA22 = []
        
        matrizB11 = []
        matrizB12 = [] 
        matrizB21 = []
        matrizB22 = []
        
        matrizResultadoParte11 = []
        matrizResultadoParte12 = []
        matrizResultadoParte21 = []
        matrizResultadoParte22 = []
        
        ayudante1 = []
        ayudante2 = []
        
        auxiliar1 = []
        auxiliar2 = []
        auxiliar3 = []
        auxiliar4 = []
        auxiliar5 = []
        auxiliar6 = []
        auxiliar7 = []

        # Asignar memoria para cada fila
        for i in range(nuevoTamanio):
            matrizA11.append([0] * nuevoTamanio)
            matrizA12.append([0] * nuevoTamanio)
            matrizA21.append([0] * nuevoTamanio)
            matrizA22.append([0] * nuevoTamanio)
            
            matrizB11.append([0] * nuevoTamanio)
            matrizB12.append([0] * nuevoTamanio)
            matrizB21.append([0] * nuevoTamanio)
            matrizB22.append([0] * nuevoTamanio)
            
            matrizResultadoParte11.append([0] * nuevoTamanio)
            matrizResultadoParte12.append([0] * nuevoTamanio)
            matrizResultadoParte21.append([0] * nuevoTamanio)
            matrizResultadoParte22.append([0] * nuevoTamanio)
            
            ayudante1.append([0] * nuevoTamanio)
            ayudante2.append([0] * nuevoTamanio)
            
            auxiliar1.append([0] * nuevoTamanio)
            auxiliar2.append([0] * nuevoTamanio)
            auxiliar3.append([0] * nuevoTamanio)
            auxiliar4.append([0] * nuevoTamanio)
            auxiliar5.append([0] * nuevoTamanio)
            auxiliar6.append([0] * nuevoTamanio)
            auxiliar7.append([0] * nuevoTamanio)
            
        # Llenamos las matrices
        for i in range(nuevoTamanio):
            for j in range(nuevoTamanio):
                matrizA11[i][j] = A[i][j]
                matrizA12[i][j] = A[i][nuevoTamanio + j]
                matrizA21[i][j] = A[nuevoTamanio + i][j]
                matrizA22[i][j] = A[nuevoTamanio + i][nuevoTamanio + j]

                matrizB11[i][j] = B[i][j]
                matrizB12[i][j] = B[i][nuevoTamanio + j]
                matrizB21[i][j] = B[nuevoTamanio + i][j]
                matrizB22[i][j] = B[nuevoTamanio + i][nuevoTamanio + j]
        
        # computing the seven aux. variables
        
       # Computando las siete variables auxiliares
        ayudante1 = [[matrizA11[i][j] + matrizA22[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]
        ayudante2 = [[matrizB11[i][j] + matrizB22[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]
        auxiliar1 = StrassenNaivStep(ayudante1, ayudante2, auxiliar1, nuevoTamanio, m)


        ayudante1 = [[matrizA21[i][j] + matrizA22[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]
        auxiliar2 = StrassenNaivStep(ayudante1, matrizB11, auxiliar2, nuevoTamanio, m)
        
        ayudante1 = [[matrizB12[i][j] - matrizB22[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]
        auxiliar3 = StrassenNaivStep(matrizA11, ayudante1, auxiliar3, nuevoTamanio, m)
        
        ayudante1 = [[matrizB21[i][j] - matrizB11[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]
        auxiliar4 = StrassenNaivStep(matrizA22, ayudante1, auxiliar4, nuevoTamanio, m)
        
        ayudante1 = [[matrizA11[i][j] + matrizA12[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]
        auxiliar5 = StrassenNaivStep(ayudante1, matrizB22, auxiliar5, nuevoTamanio, m)
        
        ayudante1 = [[matrizA21[i][j] - matrizA11[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]
        ayudante2 = [[matrizB11[i][j] + matrizB12[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]
        auxiliar6 = StrassenNaivStep(ayudante1, ayudante2, auxiliar6, nuevoTamanio, m)
        
        ayudante1 = [[matrizA12[i][j] - matrizA22[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]
        ayudante2 = [[matrizB21[i][j] + matrizB22[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]
        auxiliar7 = StrassenNaivStep(ayudante1, ayudante2, auxiliar7, nuevoTamanio, m)
        
        # Computando las cuatro partes del resultado
        matrizResultadoParte11 = [[auxiliar1[i][j] + auxiliar4[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]
        matrizResultadoParte11 = [[matrizResultadoParte11[i][j] - auxiliar5[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]
        matrizResultadoParte11 = [[matrizResultadoParte11[i][j] + auxiliar7[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]

        matrizResultadoParte12 = [[auxiliar3[i][j] + auxiliar5[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]

        matrizResultadoParte21 = [[auxiliar2[i][j] + auxiliar4[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]

        matrizResultadoParte22 = [[auxiliar1[i][j] + auxiliar3[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]
        matrizResultadoParte22 = [[matrizResultadoParte22[i][j] - auxiliar2[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]
        matrizResultadoParte22 = [[matrizResultadoParte22[i][j] + auxiliar6[i][j] for j in range(nuevoTamanio)] for i in range(nuevoTamanio)]

        # Almacenar resultados en la matriz resultado
        for i in range(nuevoTamanio):
            for j in range(nuevoTamanio):
                Result[i][j] = matrizResultadoParte11[i][j]
        
        for i in range(nuevoTamanio):
            for j in range(nuevoTamanio):
                Result[i][nuevoTamanio + j] = matrizResultadoParte12[i][j]
        
        for i in range(nuevoTamanio):
            for j in range(nuevoTamanio):
                Result[nuevoTamanio + i][j] = matrizResultadoParte21[i][j]
        
        for i in range(nuevoTamanio):
            for j in range(nuevoTamanio):
                Result[nuevoTamanio + i][nuevoTamanio + j] = matrizResultadoParte22[i][j]
        
    else:
       # matrizResultado = naiv_standard(matrizA, matrizB, matrizResultado, len(matrizA), len(matrizB), len(matrizResultado))
        Result = NaivStandard(A, B, Result, cantidadFilasMatrices, cantidadFilasMatrices, cantidadFilasMatrices)
    
    return Result